ouroboros-consensus/src/Ouroboros/Consensus/HardFork/History/Util.hs

```diff
@@ -29,7 +30,11 @@ addEpochs n (EpochNo x) = EpochNo (x + n)
 -- | @countSlots to fr@ counts the slots from @fr@ to @to@ (@to >= fr@)
 countSlots :: HasCallStack => SlotNo -> SlotNo -> Word64
 countSlots (SlotNo to) (SlotNo fr) = assert (to >= fr) $ to - fr
+  where
+    _ = keepRedundantConstraint (Proxy HasCallStack)
```

Comment:
Is this necessary? GHC never complains about redundant `HasCallStack`

Reply:
```
src/Ouroboros/Consensus/HardFork/History/Util.hs:30:1: error: [-Wredundant-constraints, -Werror=redundant-constraints]
    • Redundant constraint: HasCallStack
    • In the type signature for:
           countSlots :: HasCallStack => SlotNo -> SlotNo -> Word64
   |
30 | countSlots :: HasCallStack => SlotNo -> SlotNo -> Word64
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

src/Ouroboros/Consensus/HardFork/History/Util.hs:34:1: error: [-Wredundant-constraints, -Werror=redundant-constraints]
    • Redundant constraint: HasCallStack
    • In the type signature for:
           countEpochs :: HasCallStack => EpochNo -> EpochNo -> Word64
   |
34 | countEpochs :: HasCallStack => EpochNo -> EpochNo -> Word64
```

Reply:
Oh I see, that's because `assert` does not use the `HasCallStack`, it is only when `assert` is replaced by ghc with [`assertError`](https://downloads.haskell.org/~ghc/8.10.7/docs/html/libraries/base-4.14.3.0/GHC-IO-Exception.html#v:assertError) `HasCallStack` becomes useful.